kernel/kernel-api/src/main/java/io/delta/kernel/internal/CreateTableTransactionBuilderImpl.java

```diff
@@ -0,0 +1,137 @@
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class CreateTableTransactionBuilderImpl implements CreateTableTransactionBuilder {
+
+  private final String unresolvedPath;
+  private final StructType schema;
+  private final String engineInfo;
+
+  private Optional<Map<String, String>> tableProperties = Optional.empty();
+  private Optional<DataLayoutSpec> dataLayoutSpec = Optional.empty();
+
+  /**
+   * Number of retries for concurrent write exceptions to resolve conflicts and retry commit. In
+   * Delta-Spark, for historical reasons the number of retries is really high (10m). We are starting
+   * with a lower number by default for now. If this is not sufficient we can update it.
+   */
+  private int maxRetries = 200;
+
+  public CreateTableTransactionBuilderImpl(String tablePath, StructType schema, String engineInfo) {
+    this.unresolvedPath = requireNonNull(tablePath, "tablePath is null");
+    this.schema = requireNonNull(schema, "schema is null");
+    this.engineInfo = requireNonNull(engineInfo, "engineInfo is null");
+  }
+
+  @Override
+  public CreateTableTransactionBuilder withTableProperties(Map<String, String> properties) {
+    requireNonNull(properties, "properties cannot be null");
+    this.tableProperties =
+        Optional.of(
+            Collections.unmodifiableMap(
+                TableConfig.validateAndNormalizeDeltaProperties(properties)));
+    return this;
+  }
+
+  @Override
+  public CreateTableTransactionBuilder withDataLayoutSpec(DataLayoutSpec spec) {
+    requireNonNull(spec, "spec cannot be null");
+    this.dataLayoutSpec = Optional.of(spec);
+    return this;
+  }
+
+  @Override
+  public CreateTableTransactionBuilder withMaxRetries(int maxRetries) {
+    checkArgument(maxRetries >= 0, "maxRetries must be >= 0");
+    this.maxRetries = maxRetries;
+    return this;
+  }
+
+  @Override
+  public Transaction build(Engine engine) {
+    requireNonNull(engine, "engine cannot be null");
+    String resolvedPath = resolveTablePath(engine, unresolvedPath);
+
+    // Extract partition and clustering columns from the data layout spec
+    Optional<List<String>> partitionColumns =
+        dataLayoutSpec
+            .filter(DataLayoutSpec::hasPartitioning)
+            .map(
+                spec ->
+                    spec.getPartitionColumns().stream()
+                        .map(col -> col.getNames()[0])
+                        .collect(Collectors.toList()));
+    Optional<List<Column>> clusteringColumns =
+        dataLayoutSpec
+            .filter(DataLayoutSpec::hasClustering)
+            .map(DataLayoutSpec::getClusteringColumns);
+
+    // Build the transaction metadata using TransactionMetadataFactory
+    TransactionMetadataFactory.Output txnMetadata =
+        TransactionMetadataFactory.buildCreateTableMetadata(
+            resolvedPath,
+            schema,
+            tableProperties.orElse(Collections.emptyMap()),
+            partitionColumns,
```

Comment:
Remind me: is it worth considering having the factory take in the DataLayoutSpec instead? would other builders do similar logic to the above?

or, for example, new partition columns isn't possible on UPDATE, so that should *not* be in the common factory API?

Reply:
I think this might be shared with ReplaceTableBuilder -- but notably not with Update. I honestly think there might be a fair amount of overlap between replace/create in the underlying implementation. I'm of the opinion to leave this as is for now, and address this in the individual PR for Replace where we can see what is common code and how's the best way to factor it out. It's easier to make those decisions once we have both builders being implemented.

Reply:
SGTM 👍